### ISSUE TITLE
mail() - Unable to send email

### ISSUE THREAD

**Issue by ivan-veres** [davidalger/warden#415]
Did a fresh install Ubuntu 21.04
- docker 20.10.8, build 3967b7d
- docker-compose 1.29.2, build 5becea4c
- nameserver added 127.0.0.1
- warden - 0.11.0

Can't send emails from container. Tested on new Magneto 2 env https://docs.warden.dev/environments/magento2.html
When installed try reset admin password.

Error message is displayed "We're unable to send the password reset email."

php-fpm_1 logs
```
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='mailhog:1025'",
.
[25-Aug-2021 09:21:22] WARNING: [pool www] child 55 said into stderr: "error sending mail",
[25-Aug-2021 09:21:22] WARNING: [pool www] child 55 said into stderr: "2021/08/25 09:21:22 dial tcp 184.168.221.27:1025: getsockopt: connection timed out",

```

Mailhog logs
```
2021/08/25 08:59:04 [SMTP] Binding to address: 0.0.0.0:1025
[HTTP] Binding to address: 0.0.0.0:8025
2021/08/25 08:59:04 Serving under http://0.0.0.0:8025/
```

dig from php-fpm container
```
dig mailhog

; <<>> DiG 9.11.26-RedHat-9.11.26-4.el8_4 <<>> mailhog
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 62202
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;mailhog.			IN	A

;; ANSWER SECTION:
mailhog.		600	IN	A	172.23.0.4
```

curl from php-fpm
```
curl mailhog:1025
220 mailhog.example ESMTP MailHog
500 Unrecognised command
500 Unrecognised command
500 Unrecognised command
500 Unrecognised command
500 Unrecognised command
```

**Comment by davidalger** [comment]
I could not seem to reproduce this, so I pulled latest svc and env images, blew everything away and started back up to no avail:

1. Pulled latest svc images, blow away all docker data and re-create traefik, mailhog, etc

```
warden svc pull
warden svc down -v
warden svc up
```

2. Pull latest env images, blow away, re-create environment containers

```
warden env pull
warden env down -v
warden env up
```

3. Place this file at `webroot/pub/mail.php` for testing and request it a few times from my browser at https://app.m2sandbox.test/mail.php

```php
<?php
header('Cache-Control: private');

$message = "Line 1\r\nLine 2\r\nLine 3";

mail('customer@example.com', 'My Subject', $message, 'From: sender@example.com');
```

4. Load up mailhog interface at https://mailhog.warden.test/ and check for mail:

<img width="827" alt="image" src="https://user-images.githubusercontent.com/658281/131223069-dd09d3ee-1ad2-4429-8ec0-2bf0adbce83d.png">

Mail is sending for me both before and after updating images and re-creating all containers. Tested with latest Warden, Docker version 20.10.7, Fedora 34.

At the start of the `docker-entrypoint` script in the php-fpm container it does this:
```
cat /etc/php.d/05-additions.ini.template \
  | envsubst '${MAILHOG_HOST} ${MAILHOG_PORT}' \
  | sudo tee /etc/php.d/05-additions.ini
```

Should end up looking like this:
```
$ cat /etc/php.d/05-additions.ini
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='mailhog:1025'"
```

It almost sounds like, based on the error message, that the address written is either incorrect, not listening or if the `mailhog` hostname is correct perhaps it's resolving incorrectly to the wrong IP (i.e. something other than the IP of the docker mailhog container). Or it could be a firewall rule on linux host blocking it I suppose, but rather unlikely I would think.

The error message from two of you above has shown this bit:
```
dial tcp 184.168.221.27:1025: getsockopt: connection timed out
```

That IP looks wrong, and is likely where the problem lies, as `mailhog` in the smtp-addr should be resolving to a private IP address, not a public IP (which in this case is one that belongs to GoDaddy)

**Comment by ivan-veres** [comment]
```
==> [10:13:01] warden shell
www-data@m2-php-fpm:/var/www/html$ cat /etc/php.d/05-additions.ini.template
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='${MAILHOG_HOST}:${MAILHOG_PORT}'"

www-data@m2-php-fpm:/var/www/html$ cat /etc/php.d/05-additions.ini.template  | envsubst '${MAILHOG_HOST} ${MAILHOG_PORT}'  | sudo tee /etc/php.d/05-additions.ini
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='mailhog:1025'"

www-data@m2-php-fpm:/var/www/html$ cat /etc/php.d/05-additions.ini
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='mailhog:1025'"
```

I believe this is ok.
I'll reinstall Ubuntu 21.04 to see if problem persists, since on Fedora it's ok.

**Comment by lbajsarowicz** [comment]
If I hardcode the Mailhog IP in `/etc/php.d/05-additions.ini` everything works like a charm:
```
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='172.20.0.4:1025'"
```
But with the default value of `mailhog:1025` it does not work at all.
I think the issue appeared for the first time when I updated the way that DNS resolution is handled.

**Comment by lbajsarowicz** [comment]
If I hardcode the Mailhog IP in `/etc/php.d/05-additions.ini` everything works like a charm:
```
sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='172.20.0.4:1025'"
```
But with the default value of `mailhog:1025` it does not work at all.
I think the issue appeared for the first time when I updated the way that DNS resolution is handled.

TLDR of the issue:
```
$ warden shell
$ php -r 'mail("lukasz.bajsarowicz@spam.com", "Test","test");'
error sending mail
2021/08/31 21:36:18 dial tcp 184.168.221.27:1025: getsockopt: connection timed out
```

**Comment by ivan-veres** [comment]
Another interesting find:
```
$ ping mailhog
PING mailhog.net (184.168.221.27) 56(84) bytes of data.

--- mailhog.net ping statistics ---
8 packets transmitted, 0 received, 100% packet loss, time 7163ms

~/Projects/magento/m2 (master)
$ ws

==> [12:40:33] warden shell
www-data@m2-php-fpm:/var/www/html$ ping mailhog
PING mailhog (172.20.0.3) 56(84) bytes of data.
64 bytes from mailhog.wmg_default (172.20.0.3): icmp_seq=1 ttl=64 time=0.093 ms
64 bytes from mailhog.wmg_default (172.20.0.3): icmp_seq=2 ttl=64 time=0.124 ms
```

From container it resolves to proper IP/host but from OS it is that `184.168.221.27` IP.

```
$ cat /etc/resolvconf/resolv.conf.d/base 
search home net
nameserver 127.0.0.1
nameserver 1.1.1.1
nameserver 1.0.0.1
```

```
$ systemd-resolve --status
Global
         Protocols: -LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
  resolv.conf mode: foreign
Current DNS Server: 127.0.0.1
       DNS Servers: 127.0.0.1 103.86.96.100 103.86.99.100
        DNS Domain: home net

Link 2 (eno2)
Current Scopes: none
     Protocols: -DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported

Link 3 (wlo1)
    Current Scopes: DNS
         Protocols: +DefaultRoute +LLMNR -mDNS -DNSOverTLS DNSSEC=no/unsupported
Current DNS Server: 192.168.8.1
       DNS Servers: 192.168.8.1
....
```

**Comment by AndreiLN** [comment]
As @lbajsarowicz, worked for me on Ubuntu 21 after removing `resolvconf`.

**Comment by davidalger** [comment]
It's undoubtedly this line in resolvconf causing the issue:
```
search home net
```

The IP it's (incorrectly) resolving to is 184.168.221.27, which happens to be the IP address for `mailhog.net`:
```
nslookup mailhog.net
Server:		172.16.2.0
Address:	172.16.2.0#53

Non-authoritative answer:
Name:	mailhog.net
Address: 184.168.221.27
```

It's a little strange to me that `net` is included in the list of search domains. Does Ubuntu include this by default? On Fedora when the hostname is `foo.bar` the setting in `/etc/resolv.conf` is configured to `search bar`

**Comment by manuelcanepa** [comment]
> If I hardcode the Mailhog IP in `/etc/php.d/05-additions.ini` everything works like a charm:
> 
> ```
> sendmail_path = "/usr/local/bin/mhsendmail --smtp-addr='172.20.0.4:1025'"
> ```
> 
> But with the default value of `mailhog:1025` it does not work at all. I think the issue appeared for the first time when I updated the way that DNS resolution is handled.
> 
> TLDR of the issue:
> 
> ```
> $ warden shell
> $ php -r 'mail("lukasz.bajsarowicz@spam.com", "Test","test");'
> error sending mail
> 2021/08/31 21:36:18 dial tcp 184.168.221.27:1025: getsockopt: connection timed out
> ```
> 
> When I uninstalled `resolvconf` on Ubuntu, everything works correctly o.O

Same here, is working fine after remove `resolvconf`
